docs/registry/README.md

```diff
@@ -0,0 +1,193 @@
+
+Our design revolves around the following core requirements:
+
+1. We should have a Registry per namespace to enforce isolation.
+2. The Registry should contain the event types that can be consumed from
+the eventing mesh. If an event type is not ready for consumption, we should explicitly indicate so.
+3. The event types stored in the Registry should contain all the required information 
+for a consumer to create a Trigger without resorting to some other OOB mechanism.
+
+## Design Ideas
+
+### EventType CRD
+
+We propose having a namespaced EventType CRD. Here is an example of how a CR would look like:
+
+```yaml
+apiVersion: eventing.knative.dev/v1alpha1
+kind: EventType
+metadata:
+  name: repopush
+spec:
+  type: repo:push
+  source: my-user/my-repo
+  schema: my-schema
+  broker: default
+```
+
+
+- The `name` of the EventType is advisory, non-authoritative. Given that Cloud Event types can 
+contain characters that may not comply with Kubernetes naming conventions, we will (slightly) 
+modify those names to make them K8s-compliant, whenever we need to generate them. 
+
+- `type` is authoritative. This refers to the Cloud Event type as it enters into the eventing mesh. 
+
+- `source`: an identifier of where we receive the event from. This might not necessarily be the Cloud Event source 
+attribute. If we receive the event from our receive adaptors, the info might come in a Cloud Event custom extension (e.g., from).
+
+- `schema` is a URI with the EventType schema. It may be a JSON schema, a protobuf schema, etc. It is optional.
+
+- `broker` refers to the Broker that can provide the EventType. 
+
+### EventType Instantiation
+
+We foresee the following ways of populating the Registry:
+
+**1. Event Source CR installation**
+
+Upon installation of an Event Source CR, the source will register its EventTypes.
+
+Example:
+
+```yaml
+apiVersion: sources.eventing.knative.dev/v1alpha1
+kind: GitHubSource
+metadata:
+  name: github-source-sample
+  namespace: default
+spec:
+  eventTypes:
+    - push
+    - pull_request
+  ownerAndRepository: my-other-user/my-other-repo
+  accessToken:
+    secretKeyRef:
+      name: github-secret
+      key: accessToken
+  secretToken:
+    secretKeyRef:
+      name: github-secret
+      key: secretToken
+  sink:
+    apiVersion: eventing.knative.dev/v1alpha1
+    kind: Broker
+    name: default
+
+```
+ 
+By applying this file, two EventTypes will be registered, with types `push` and `pull_request`, 
+source `my-other-user/my-other-repo`, for the `default` Broker in the `default` namespace.
+
+**2. Manual User Registration**
+
+A user manually `kubectl applies` an EventType CR.
+
+Example: 
+
+```yaml
+apiVersion: eventing.knative.dev/v1alpha1
+kind: EventType
+metadata:
+  name: repofork
+  namespace: default
+spec:
+  type: repo:fork
+  source: my-other-user/my-other-repo
+  broker: dev
+``` 
+
+This would register the EventType named `repofork` with type `repo:fork`, source `my-other-user/my-other-repo` 
```

Comment:
I see `my-other-user/my-other-repo` inside the GH source, but not as name .. am I missing something ?

Reply:
Not as source you mean? 

As the cloud event source field is somewhat useless. For example, the current GH adapter for a push request sends as cloud event source `https://github.com/<user>/<repo>/pull/<pull_id>`, there is no way to do exact matching on that on a trigger. 
I propose to augment the cloud event in adapters with some custom extension that will allow us to do better filtering. In the case of GH, have that user/repo, in the case of kafka, maybe just the consumerGroup/topicName, and so on. 
I know this is not in this README as of now, I will add these details for further clarification. Thanks for pointing that out.
In case you want to take a look at some initial work that I did, you can take a look at https://github.com/knative/eventing-sources/pull/300

Reply:
@deissnerk I meant that the Ce-Source is too specific and can contain information only known at runtime. For example, that pull_id, or some commit_id, or whatever. 
If a downstream consumer wants to consume events from a particular Ce-Source, then in our current model the consumer needs to create a Trigger filtering for events matching that Ce-Source. As of now, we only support exact matching on Ce-Source, which IMHO is somewhat useless. With more advanced filtering mechanisms we should be OK. 
But even then, if we rely on the cloud event source to uniquely identify EventTypes, e.g., if we use (type, source, schema) as "primary key" for EventTypes, then the registry will blow up in the sense that we will be adding a bunch of identical EventTypes... For example (omitting irrelevant fields):
```
type: dev.knative.source.github.pull_request
source: https://github.com/nachocano/eventing/pull/1

type: dev.knative.source.github.pull_request
source: https://github.com/nachocano/eventing/pull/2

...

type: dev.knative.source.github.pull_request
source: https://github.com/nachocano/eventing/pull/100
```
Does it make sense?

Yes, @n3wscott mentioned this new `subject` field they are proposing in the spec. I think it will be much more useful than `source`, but not sure when it will be available.

Reply:
@nachocano The subject field was approved yesterday. So it should be there in the upcoming CE version.

Reply:
@deissnerk thanks for the heads up!